common/djangoapps/student/roles.py

```diff
@@ -69,12 +77,20 @@ def __init__(self, user):
                 CourseAccessRole.objects.filter(user=user).all()
             )
 
+    @staticmethod
+    def _get_roles(role):
+        """
+        Return the roles that should have the same permissions as the specified role.
+        """
+        return ACCESS_ROLES_INHERITANCE.get(role, set()) | {role}
+
     def has_role(self, role, course_id, org):
         """
-        Return whether this RoleCache contains a role with the specified role, course_id, and org
+        Return whether this RoleCache contains a role with the specified role
+        or a role that inherits from the specified role, course_id and org.
         """
         return any(
-            access_role.role == role and
+            access_role.role in self._get_roles(role) and
```

Comment:
@0x29a, I found a small bug here. If we go to the Instructor Dashboard, add a user as `Limited Staff`, and then try to add it as `Staff`, nothing will happen. That's because we're doing a role check to avoid duplicates. To fix this, we can change the following part https://github.com/openedx/edx-platform/blob/252c88e041a4a3733b72c0f5778ceba514c89703/common/djangoapps/student/roles.py#L205-L207

to something like this:
```python
            if user.is_authenticated and user.is_active:
                CourseAccessRole.objects.get_or_create(
                    user=user, role=self._role_name, course_id=self.course_key, org=self.org
                )
```

Alternatively, we could add a `strict_checking` argument to  `has_user` and `has_role` methods. This way, we could ignore the `ACCESS_ROLES_INHERITANCE` in this scenario. The first approach sounds more straightforward, though.

Reply:
Ah, thanks for catching this, @Agrendalath. I'll add a test for this.

Reply:
Upd: fixed, but there seem to be no good place to add a test for this.